### PR TITLE
Add sudo option for intel_rdt plugin

### LINKED ISSUE #9380: When running as non-root user, intel_rdt plugin has no output
### Relevant telegraf.conf:

```toml
[[inputs.intel_rdt]]
cores = ["0-5,8-13"]
```

### System info:

Telegraf: 1.18.3+git-6ef3e51f~202105250953~ubuntu18.04.1 from  http://ppa.launchpad.net/telegraf-devs/ppa/ubuntu
Ubuntu Bionic, with kernel 5.4 (HWE)
intel-cmt-cat 4.1 from source (Bionic's package is older)

### Steps to reproduce:

1. configure the intel_rdt plugin per documentation
2. Ubuntu packages run telegraf as user 'telegraf', not as root

### Expected behavior:

Expected metrics, such as:

```
  rdt_metric{cores="0,1,2,3,4,5,8,9,10,11,12,13",host="controller:ubuntu-1",juju_application="ubuntu",juju_model="controller",juju_unit="ubuntu-1",name="IPC"} 0.88
  rdt_metric{cores="0,1,2,3,4,5,8,9,10,11,12,13",host="controller:ubuntu-1",juju_application="ubuntu",juju_model="controller",juju_unit="ubuntu-1",name="LLC"} 9984
  rdt_metric{cores="0,1,2,3,4,5,8,9,10,11,12,13",host="controller:ubuntu-1",juju_application="ubuntu",juju_model="controller",juju_unit="ubuntu-1",name="LLC_Misses"} 305356
  rdt_metric{cores="0,1,2,3,4,5,8,9,10,11,12,13",host="controller:ubuntu-1",juju_application="ubuntu",juju_model="controller",juju_unit="ubuntu-1",name="MBL"} 18.9
  rdt_metric{cores="0,1,2,3,4,5,8,9,10,11,12,13",host="controller:ubuntu-1",juju_application="ubuntu",juju_model="controller",juju_unit="ubuntu-1",name="MBR"} 0.3
  rdt_metric{cores="0,1,2,3,4,5,8,9,10,11,12,13",host="controller:ubuntu-1",juju_application="ubuntu",juju_model="controller",juju_unit="ubuntu-1",name="MBT"} 19.
```

### Actual behavior:

pqos would need to have priv escalation (sudo?) as part of the plugin in order for telegraf to run as anything other than the root user, if this plugin is to work.  Currently we see no data in metrics from the intel_rdt plugin, logs show that pqos fails:

```
2021-06-16T03:42:13Z E! [inputs.intel_rdt] Error: pqos: exit status 1
2021-06-16T03:42:13Z E! [inputs.intel_rdt] failed to shut down pqos: os: process already finished
```

### Additional info:

Running telegraf in the foreground as root works perfectly - this issue is purely that due to security requirements telegraf is not to run as root.

### PR DESCRIPTION
**Pull request by guoqiao**
### Required for all PRs:

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.

resolves #9380

When running telegraf as a non-root user, this change allows the use
of sudo to run the necessary pqos command for the intel_rdt plugin.

### CODE REVIEW

**Comment by guoqiao** [comment]
!signed-cla

**Comment by guoqiao** [comment]
This PR is created based on and to supersede #9465  since @xavpaice  is on leave and task is assigned to me.
Changes:
- rename option `sudo` to `use_sudo`,  to be consistent with existing 10+ plugins(search `sudo` in `etc/telegraf.conf` to see them)
- revert irrelevant trailing white space change
- rm TODO comment, as my understanding, the code in `func validatePqosPath` are still valid with sudo, so nothing to change.

Review appreciated :)

**Review by srebhan — APPROVED** [review]
Looks good to me. Thanks for tackling this issue @guoqiao!

**Comment by guoqiao** [comment]
@srebhan   thanks for the approval!  Is there anything else I need to do, to get it merged to master, please ?

**Comment by KubaTrojan** [comment]
@guoqiao could you refer to [this comment](https://github.com/influxdata/telegraf/pull/9465#issuecomment-874158029)?

**Comment by srebhan** [comment]
@KubaTrojan can you please just copy-paste your comment here so we have everything in one place.

**Comment by srebhan** [comment]
@guoqiao the usual procedure is that we get an additional review (from Influx people) before merging. The merging time also depends on the release cycle as this change will probably not qualify for a bug-fix release. :-)
Can you please also check the comment by @KubaTrojan as it seems operation has some problems... Will wait with applying the "ready for final review" tag until this issue is solved.

**Comment by KubaTrojan** [comment]
As @srebhan asked, I copy [my comment](https://github.com/influxdata/telegraf/pull/9465#issuecomment-874158029) here:

I fetched your changes and I got some problems with closing telegraf. Pqos tool could be set to working with sudo (by adding it to sudoers file) and be successfully started by unprivileged telegraf plugin, but there is an issue with closing its work. The plugin freezes when ending work (started with: ```telegraf --test-wait 5 --config=telegraf.conf```) while config param ```sudo=true```, because plugin itself, cannot end pqos root process.

Is this issue exists on your side also?
 
The proper pqos tool closing is important, because the tool needs to be closed by sending an interrupt (SIGINT, -2) signal to it, due to cleaning it does in _/sys/fs/resctrl/mon_groups_ directory. But seems that unprivileged telegraf plugin cannot send an interrupt signal to a privileged pqos process. However, if the command starts with ```bin/sh -c sudo -n pqos ...```, it can be terminated, by terminating ```bin/sh -c``` as the parent process of ```sudo -n pqos ...```. But it doesn't fix the issue of leaving garbage in _/sys/fs_ directory.

One option i thought is to make a custom script with the capability to close a specific pqos process by finding its pid, and to add that script to sudoers file also. The script could look like this:

```
#!/bin/bash

PID=$(pgrep -f 'sudo -n path/to/pqos -r --iface-os --mon-file-type=csv')
sudo kill -2 "-$PID"
```
Where ```sudo -n path/to/pqos -r --iface-os --mon-file-type=csv``` is the specific command which was used to run pqos. It may be prefixed with ```bin/sh -c``` or point to pqos wrapper script.

 
And in code, telegraf plugin which was started as non-root, should have additional functionality to run the closing script with ```sudo``` while stopping.

**Comment by guoqiao** [comment]
@srebhan  @KubaTrojan 
I did see that comment,  I thought xav chosed to use the `bin/sh -c sudo ....` solution and we don't care too much about `leaving garbage in /sys/fs directory` issue, for now.

I am trying to build a ppa with this patch and deploy in a test env, I will see how that goes and come back to update here.

Thank you both.

**Comment by srebhan** [comment]
I think @p-zak should take a look at this as his knowledge on the topic is much more profound.

**Comment by guoqiao** [comment]
I have built a ppa with 1.18.3 + this patch:  [ppa:guoqiao/telegraf](https://launchpad.net/~guoqiao/+archive/ubuntu/telegraf) and deploy it to a host with rdt support.  

1) With the `/bin/sh -c sudo /path/to/pqos ...` cmd and `use_sudo = true`,  rdt plugin seems stop/start fine.
2) Re:  garbage in /sys/fs directory,  I didn't see it as a major issue for our current requirement.

For now, we will use above ppa in our cloud, until this patch is landed.

**Comment by xavpaice** [comment]
I tried using a killall with the signal which worked perfectly, and have a
small patch in progress but am away from work at the moment, maybe if this
patch is merged/closed we can raise a separate bug and patch to fix the
issue closing the plugin.

**Comment by p-zak** [comment]
> I think @p-zak should take a look at this as his knowledge on the topic is much more profound.

@p-zak is on his holidays right now :)
@KubaTrojan is an original author of this plugin thus he has the best knowledge what problems proposed change can introduce. Did we find solution in which pqos tool is closed properly?

**Comment by guoqiao** [comment]
> > I think @p-zak should take a look at this as his knowledge on the topic is much more profound.
> 
> @p-zak is on his holidays right now :)
> @KubaTrojan is an original author of this plugin thus he has the best knowledge what problems proposed change can introduce. Did we find solution in which pqos tool is closed properly?

I think so, with `/bin/sh -s sudo ....`,  as I can see, pqos can stop / start without issue.

**Comment by xavpaice** [comment]
The process itself can be killed but without the correct signal it leaves
trash behind that needs cleaning. Using sudo killall-s 1 pqos works but we
need to add that if the process is started with sudo.

**Comment by KubaTrojan** [comment]
@xavpaice so the situation where pqos still works, even after telegraf stops, occurs for you?

@guoqiao 
Let me show in code, where the problem exists on my system:
https://github.com/influxdata/telegraf/blob/3aef3c418f36e3cae9c7d9210dbaad2f84e2c002/plugins/inputs/intel_rdt/intel_rdt.go#L347-L358
Above, in line 349 the plugin tries to send an interrupt signal to previously started process (in this case `/bin/sh -s sudo -n pqos ....`). `Signal()` method, returns `nil` but pqos is still running. It is because the signal was successfully sent to unprivileged `/bin/sh` command. However its child process: `sudo -n pqos ...` belongs to root, so no effect. Thereby the plugin will not even try to kill desired process and unfortunately stuck in this place on `cmd.Wait()`:
https://github.com/influxdata/telegraf/blob/3aef3c418f36e3cae9c7d9210dbaad2f84e2c002/plugins/inputs/intel_rdt/intel_rdt.go#L295-L298

What I could propose, is to change `shutDownPqos` method, to check after sending interrupt signal, if process is still running. It could be checked like this:
```
// check if pqos process still exists
if err := pqos.Process.Signal(syscall.Signal(0)); err == os.ErrProcessDone {
	return nil
}
```
And if the command process still works, it should be terminated/killed. (unfortunately without cleaning. Maybe a proper warning message should be logged?)

Also, I think below line should be changed, as it is potential place where the plugin may freeze:
https://github.com/influxdata/telegraf/blob/3aef3c418f36e3cae9c7d9210dbaad2f84e2c002/plugins/inputs/intel_rdt/intel_rdt.go#L297

It can be simplified to just: `r.Log.Errorf("pqos: %v", err)` (In line 292 also?).

What do you think about it?

**Comment by xavpaice** [comment]
FWIW, running this plugin from a PPA built from the change here:

* `systemctl stop telegraf` takes 90s to timeout,  but does appear to stop the pqos executable (just not using the interrupt signal).
* The pqos executable does stop and clear from the process list (presumably because it's parent process was terminated)
* `/sys/fs/resctrl/mon_groups/pqos-*` is left behind from the last running pqos pid, however the files are empty and not updated once pqos is stopped.  The lack of cleanup appears to be only cosmetic but it isn't ideal.

I thought about adding a test in `shutDownPqos` to determine if the command is running under sudo, and if so, use `killall -SIGINT pqos` to stop the process rather than the `pqos.Process.Signal(os.Interrupt)` otherwise.

e.g. something I had drafted earlier but not tested out:

```
 func shutDownPqos(pqos *exec.Cmd) error {
        if pqos.Process != nil {
-               err := pqos.Process.Signal(os.Interrupt)
-               if err != nil {
-                       err = pqos.Process.Kill()
+               if pqos.Path == "/bin/sh" {  // the process is running under sudo
+                       cmd := exec.Command("sudo", "killall", "-SIGINT", "pqos")
+                       err := cmd.Start()
+                       if err == nil {
+                               err = cmd.Wait()
+                       }
+                       // TODO find a way to make sure it's dead
                        if err != nil {
                                return fmt.Errorf("failed to shut down pqos: %v", err)
                        }
+               } else {
+                       err := pqos.Process.Signal(os.Interrupt)
+                       if err != nil {
+                               err = pqos.Process.Kill()
+                               if err != nil {
+                                       return fmt.Errorf("failed to shut down pqos: %v", err)
+                               }
+                       }
                }
        }
        return nil

```

**Comment by helenosheaa** [comment]
Looks like we've all moved over to #9527, shall we close this PR?

**Comment by guoqiao** [comment]
> Looks like we've all moved over to #9527, shall we close this PR?

Sure, I will close it now. Thanks!

**Pull request closed without merging**
